service/kernel/verbose_mode.h

```diff
@@ -0,0 +1,294 @@
+    template <typename ValPtr>
+    auto put(const char * const key, const ValPtr p) ->
+        typename std::enable_if<std::is_pointer<typename std::decay<ValPtr>::type>::value, json &>::type
+    {
+        comma_if_needed();
+        need_comma = false;
+        write('"');
+        write(key);
+        write("\":");
+
+        need_comma = false;
+        begin();
+        json_print(*this, p);
+        end();
+
+        need_comma = true;
+        return *this;
+    }
+
+    json & put(const char * const key, const obj_begin_t &);
+
+    json & put(const obj_end_t &);
+
+    void finalize();
+
+    ~json();
+
+private:
+    // raw output
+    static void write(const char * const str);
+    static void write(const char c);
+    static void write(const int i);
+    static void write(const unsigned int u);
+    static void write(const unsigned long u);
+    static void write(const long long int i);
+    static void write(const unsigned long long int u);
+    static void write(const double d);
+
+    void comma_if_needed();
+
+    void begin();
+
+    void end();
+
+    // todo: mv2cpp
+    void write_escape(const char * const str);
+
+    template <typename ValPtr>
+    friend auto json_print(json & writer, const ValPtr p) -> typename std::enable_if<std::is_pointer<typename std::decay<ValPtr>::type>::value>::type;
+
+    int depth;
+    bool need_comma;
+};
+
+// pointer dispatcher
+template <typename ValPtr>
+auto json_print(json & writer, const ValPtr p) -> typename std::enable_if<std::is_pointer<typename std::decay<ValPtr>::type>::value>::type
+{
+    if (p)
+    {
+        json_print(writer, *p);
+    }
+    else
+    {
+        writer.write("\"ptr\":\"nullptr\"");
+    }
+}
+
+// mv2cpp or use inline to break ODR
+// algorithms::kmeans::Parameter &
+void json_print(json & writer, const algorithms::kmeans::Parameter & val);
+
+// data_management::NumericTable &
+// todo: mv2cpp
+void json_print(json & writer, const data_management::NumericTable & val);
+
+void json_print(json & writer, ...);
+
+template <typename algorithmFPType>
+constexpr const char * fpTypeToStr()
+{
+    return "unknown type";
+}
+
+template <>
+constexpr const char * fpTypeToStr<float>()
+{
+    return "float";
+}
+
+template <>
+constexpr const char * fpTypeToStr<double>()
+{
+    return "double";
+}
+
+template <>
+constexpr const char * fpTypeToStr<int>()
+{
+    return "int";
+}
+
+template <typename algorithmFPType, CpuType cpu>
+struct kernel_verbose_raii
+{
+    template <typename... Args>
+    kernel_verbose_raii(const char * const file, Args... args) : file_name(file)
+    {
+        if (verbose_t::level == 2)
+        {
+            writer.put("kernel file", file_name).put("algorithmFPType", fpTypeToStr<algorithmFPType>());
+            writer.put("env", begin).put("cpu", cpuTypeToStr(cpu)).put(end);
+            writer.put("args", begin);
+            put(args...);
+            writer.put(end);
```

Comment:
Structure of json are defined here;
`put("env", begin)` - to start nested object